### ISSUE TITLE
Reductions at axis=N inside lists that are empty at axis=N-1 at the end of the N-1 array.

### ISSUE THREAD

**Issue by nikoladze** [scikit-hep/awkward#264]
First, thanks for this awesome package! I'm looking forward to the new ways of analysis that we will be able to do with it.

I might have found a problem with `ak.any` and `axis=2` which seems to happen in cases when the last element of the array is an empty list:

```pycon
>>> import awkward1 as ak
>>> a = ak.from_iter([[], [[False, False, True]], [], [[False], [True, False]], []])
>>> ak.any(a, axis=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nikolai/venv/akward/lib/python3.8/site-packages/awkward1/highlevel.py", line 1108, in __repr__
    value = awkward1._util.minimally_touching_string(limit_value,
  File "/home/nikolai/venv/akward/lib/python3.8/site-packages/awkward1/_util.py", line 1166, in minimally_touching_string
    r = next(rightgen)
  File "/home/nikolai/venv/akward/lib/python3.8/site-packages/awkward1/_util.py", line 1145, in forever
    for token in iterable:
  File "/home/nikolai/venv/akward/lib/python3.8/site-packages/awkward1/_util.py", line 1107, in backward
    for token in backward(x[i], sp):
ValueError: in ListOffsetArray64 attempting to get 4, offsets[i] > offsets[i + 1]
>>> ak.any(a[:-1], axis=2)
<Array [[], [True], [], [False, True]] type='4 * var * bool'>
```
with `a[:-2]` it sometimes also raises the Exception, but sometimes produces an incorrect result:

```pycon
>>> ak.any(a[:-2], axis=2)
<Array [[], [], [True]] type='3 * var * bool'>
```

I tried this on the current master (0.2.20)

**Comment by jpivarski** [comment]
Aha! It's because some of the lists it's trying to reduce over don't exist at the level that you're looking (you have some empty lists at `axis=1` and are reducing over `axis=2`). This is indeed a bug.

Furthermore, it only happens if the empty list at `axis=1` is _last_. Here's a simpler example (I use prime numbers and `ak.prod` because it's easier to see what it's grouping; booleans are hard to follow by eye).

```python
>>> ak.prod(ak.Array([[[2, 3, 5]], [[7], [11]], [[]]]), axis=-1)
<Array [[30], [7, 11], [1]] type='3 * var * int64'>
>>> ak.prod(ak.Array([[[2, 3, 5]], [[7], [11]], []]), axis=-1)
Traceback (most recent call last):
...
ValueError: in ListOffsetArray64 attempting to get 2, offsets[i] > offsets[i + 1]
```

but

```python
>>> ak.prod(ak.Array([[[2, 3, 5]], [], [[7], [11]]]), axis=-1)
<Array [[30], [], [7, 11]] type='3 * var * int64'>
```

is okay and

```python
>>> ak.prod(ak.Array([[], [[2, 3, 5]], [[7], [11]]]), axis=-1)
<Array [[], [30], [7, 11]] type='3 * var * int64'>
```

is okay. I'll look into it.

**Comment by jpivarski** [comment]
Trimming the offending last element makes it work again, so it's not a problem that the empty list is somewhere in memory; it's a problem that it's in the range that's considered:

```python
>>> ak.prod(ak.Array([[[2, 3, 5]], [[7], [11]], []])[:-1], axis=-1)
<Array [[30], [7, 11]] type='2 * var * int64'>
```

In your example, trimming the last two gets you back to having an empty `axis=1` list last. If, in your example, you only trim off enough to leave the last `axis=1` list nonempty, it works:

```python
>>> ak.any(a[:-1], axis=2)
<Array [[], [True], [], [False, True]] type='4 * var * bool'>
```
which, as I read it, is the right answer.

It looks like only one bug: the last-empty bug.

**Comment by jpivarski** [comment]
Better yet, the exception does not occur in the reduction step; it happens in the print-out because the reduction step produced invalid output. We can inspect these by looking at the layouts:

```python
>>> ak.prod(ak.Array([[[2, 3, 5]], [[7], [11]], [[]]]), axis=-1).layout
<ListOffsetArray64>
    <offsets><Index64 i="[0 1 3 4]" offset="0" length="4" at="0x563bfcc0fd00"/></offsets>
    <content><NumpyArray format="l" shape="4" data="30 7 11 1" at="0x563bfcc058d0"/></content>
</ListOffsetArray64>
```

for the good one and

```python
>>> ak.prod(ak.Array([[[2, 3, 5]], [[7], [11]], []]), axis=-1).layout
<ListOffsetArray64>
    <offsets><Index64 i="[0 1 0 3]" offset="0" length="4" at="0x563bfcbd8150"/></offsets>
    <content><NumpyArray format="l" shape="3" data="30 7 11" at="0x563bfcb2dce0"/></content>
</ListOffsetArray64>
```

for the bad one. The reduction step made the wrong offsets. It should be `[0 1 3 3]`.

**Comment by jpivarski** [comment]
It's fixed in PR #265 and I'll merge with master soon. A few more things are going to go into the next version before I deploy, however.

For your original example,

```python
>>> import awkward1 as ak
>>> a = ak.from_iter([[], [[False, False, True]], [], [[False], [True, False]], []])
>>> ak.any(a, axis=2)
<Array [[], [True], [], [False, True], []] type='5 * var * bool'>
>>> ak.any(a[:-1], axis=2)
<Array [[], [True], [], [False, True]] type='4 * var * bool'>
>>> ak.any(a[:-2], axis=2)
<Array [[], [True], []] type='3 * var * bool'>
```
